File: packages/open-memory/packages/openmemory-py/src/openmemory/temporal_graph/store.py
import time
import uuid
import json
import logging
from typing import List, Dict, Any, Optional

from ..core.db import q, db, transaction

logger = logging.getLogger("temporal")

async def insert_fact(subject: str, predicate: str, subject_object: str, valid_from: int = None, confidence: float = 1.0, metadata: Dict[str, Any] = None, user_id: Optional[str] = None) -> str:
    fact_id = str(uuid.uuid4())
    now = int(time.time() * 1000)
    valid_from_ts = valid_from if valid_from is not None else now
    
    existing_sql = "SELECT id, valid_from FROM temporal_facts WHERE subject=? AND predicate=? AND valid_to IS NULL"
    existing_params = [subject, predicate]
    if user_id:
        existing_sql += " AND user_id=?"
        existing_params.append(user_id)
    existing_sql += " ORDER BY valid_from DESC"
    existing = db.fetchall(existing_sql, tuple(existing_params))

    for old in existing:
        if old["valid_from"] < valid_from_ts:
            db.execute("UPDATE temporal_facts SET valid_to=? WHERE id=?", (valid_from_ts - 1, old["id"]))

    meta_json = json.dumps(metadata) if metadata else None

    db.execute("INSERT INTO temporal_facts(id, user_id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata) VALUES (?,?,?,?,?,?,NULL,?,?,?)",
               (fact_id, user_id, subject, predicate, subject_object, valid_from_ts, confidence, now, meta_json))

    db.commit()
    return fact_id

async def update_fact(fact_id: str, confidence: Optional[float] = None, metadata: Optional[Dict[str, Any]] = None):
    updates = []
    params = []

    if confidence is not None:
        updates.append("confidence=?")
        params.append(confidence)

    if metadata is not None:
        updates.append("metadata=?")
        params.append(json.dumps(metadata))

    if not updates: return

    updates.append("last_updated=?")
    params.append(int(time.time() * 1000))
    params.append(fact_id)

    sql = f"UPDATE temporal_facts SET {', '.join(updates)} WHERE id=?"
    db.execute(sql, tuple(params))
    db.commit()

async def invalidate_fact(fact_id: str, valid_to: int = None):
    ts = valid_to if valid_to is not None else int(time.time() * 1000)
    db.execute("UPDATE temporal_facts SET valid_to=?, last_updated=? WHERE id=?", (ts, int(time.time() * 1000), fact_id))
    db.commit()

async def delete_fact(fact_id: str):
    db.execute("DELETE FROM temporal_facts WHERE id=?", (fact_id,))
    db.commit()

async def insert_edge(source_id: str, target_id: str, relation_type: str, valid_from: int = None, weight: float = 1.0, metadata: Dict[str, Any] = None) -> str:
    edge_id = str(uuid.uuid4())
    now = int(time.time() * 1000)
    valid_from_ts = valid_from if valid_from is not None else now
    meta_json = json.dumps(metadata) if metadata else None

    db.execute("INSERT INTO temporal_edges(id, source_id, target_id, relation_type, valid_from, valid_to, weight, metadata) VALUES (?,?,?,?,?,NULL,?,?)",
               (edge_id, source_id, target_id, relation_type, valid_from_ts, weight, meta_json))
    db.commit()
    return edge_id

async def invalidate_edge(edge_id: str, valid_to: int = None):
    ts = valid_to if valid_to is not None else int(time.time() * 1000)
    db.execute("UPDATE temporal_edges SET valid_to=? WHERE id=?", (ts, edge_id))
    db.commit()

async def batch_insert_facts(facts: List[Dict[str, Any]]) -> List[str]:
    ids = []
    try:
        db.execute("BEGIN")

        now = int(time.time()*1000)

        for f in facts:
            fid = str(uuid.uuid4())
            sub, pred, obj = f["subject"], f["predicate"], f["object"]
            vf = f.get("valid_from", now)
            conf = f.get("confidence", 1.0)
            meta = f.get("metadata")
            existing = db.conn.execute("SELECT id, valid_from FROM temporal_facts WHERE subject=? AND predicate=? AND valid_to IS NULL", (sub, pred)).fetchall()
            for old in existing:
                 if old["valid_from"] < vf:
                     db.conn.execute("UPDATE temporal_facts SET valid_to=? WHERE id=?", (vf - 1, old["id"]))

            db.conn.execute("INSERT INTO temporal_facts(id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata) VALUES (?,?,?,?,?,NULL,?,?,?)",
               (fid, sub, pred, obj, vf, conf, now, json.dumps(meta) if meta else None))
            ids.append(fid)

        db.execute("COMMIT")
        return ids
    except Exception as e:
        db.execute("ROLLBACK")
        raise e

async def apply_confidence_decay(decay_rate: float = 0.01) -> int:
    now = int(time.time() * 1000)
    one_day = 86400000
    sql = """
        UPDATE temporal_facts
        SET confidence = MAX(0.1, confidence * (1 - ? * ((? - valid_from) / ?)))
        WHERE valid_to IS NULL AND confidence > 0.1
    """
    db.execute(sql, (decay_rate, now, one_day))
    db.commit()
    return db.conn.total_changes

File: packages/open-memory/packages/openmemory-js/src/temporal_graph/store.ts
import { run_async, get_async, all_async } from "../core/db";
import { env } from "../core/cfg";
import { TemporalFact, TemporalEdge } from "./types";
import { randomUUID } from "crypto";

const is_pg = env.metadata_backend === "postgres";

export interface InsertFactOptions {
    subject: string;
    predicate: string;
    object: string;
    valid_from?: Date;
    confidence?: number;
    metadata?: Record<string, any>;
    user_id?: string;
    project_id?: string;
}

export async function insert_fact(opts: InsertFactOptions): Promise<string>;
export async function insert_fact(
    subject: string,
    predicate: string,
    object: string,
    valid_from?: Date,
    confidence?: number,
    metadata?: Record<string, any>,
    user_id?: string,
    project_id?: string,
): Promise<string>;
export async function insert_fact(
    subject_or_opts: string | InsertFactOptions,
    predicate?: string,
    object?: string,
    valid_from?: Date,
    confidence?: number,
    metadata?: Record<string, any>,
    user_id?: string,
    project_id?: string,
): Promise<string> {
    // Normalize options-bag form to the positional locals used by the
    // existing implementation.
    if (typeof subject_or_opts === "object" && subject_or_opts !== null) {
        const opts = subject_or_opts;
        return _insert_fact_impl(
            opts.subject,
            opts.predicate,
            opts.object,
            opts.valid_from ?? new Date(),
            opts.confidence ?? 1.0,
            opts.metadata,
            opts.user_id,
            opts.project_id,
        );
    }
    return _insert_fact_impl(
        subject_or_opts,
        predicate as string,
        object as string,
        valid_from ?? new Date(),
        confidence ?? 1.0,
        metadata,
        user_id,
        project_id,
    );
}

const _insert_fact_impl = async (
    subject: string,
    predicate: string,
    object: string,
    valid_from: Date = new Date(),
    confidence: number = 1.0,
    metadata?: Record<string, any>,
    user_id?: string,
    project_id?: string,
): Promise<string> => {
    const id = randomUUID();
    const now = Date.now();
    const valid_from_ts = valid_from.getTime();

    const existing = await all_async(
        `
        SELECT id, valid_from FROM temporal_facts
        WHERE subject = ? AND predicate = ? AND valid_to IS NULL${user_id ? " AND user_id = ?" : ""}${project_id ? " AND (project_id = ? OR project_id = 'system_global' OR project_id IS NULL)" : ""}
        ORDER BY valid_from DESC
    `,
        [
            subject,
            predicate,
            ...(user_id ? [user_id] : []),
            ...(project_id ? [project_id] : []),
        ],
    );

    for (const old of existing) {
        if (old.valid_from < valid_from_ts) {
            await run_async(
                `UPDATE temporal_facts SET valid_to = ? WHERE id = ?`,
                [valid_from_ts - 1, old.id],
            );
            console.error(
                `[TEMPORAL] Closed fact ${old.id} at ${new Date(valid_from_ts - 1).toISOString()}`,
            );
        }
    }

    await run_async(
        `
        INSERT INTO temporal_facts (id, user_id, project_id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata)
        VALUES (?, ?, ?, ?, ?, ?, ?, NULL, ?, ?, ?)
    `,
        [
            id,
            user_id || null,
            project_id || null,
            subject,
            predicate,
            object,
            valid_from_ts,
            confidence,
            now,
            metadata ? JSON.stringify(metadata) : null,
        ],
    );

    console.error(
        `[TEMPORAL] Inserted fact: ${subject} ${predicate} ${object} (from ${valid_from.toISOString()}, confidence=${confidence}${user_id ? `, user=${user_id}` : ""}${project_id ? `, project=${project_id}` : ""})`,
    );
    return id;
};

export const update_fact = async (
    id: string,
    confidence?: number,
    metadata?: Record<string, any>,
): Promise<void> => {
    const updates: string[] = [];
    const params: any[] = [];

    if (confidence !== undefined) {
        updates.push("confidence = ?");
        params.push(confidence);
    }

    if (metadata !== undefined) {
        updates.push("metadata = ?");
        params.push(JSON.stringify(metadata));
    }

    updates.push("last_updated = ?");
    params.push(Date.now());

    params.push(id);

    if (updates.length > 0) {
        await run_async(
            `UPDATE temporal_facts SET ${updates.join(", ")} WHERE id = ?`,
            params,
        );
        console.error(`[TEMPORAL] Updated fact ${id}`);
    }
};

export const invalidate_fact = async (
    id: string,
    valid_to: Date = new Date(),
): Promise<void> => {
    await run_async(
        `UPDATE temporal_facts SET valid_to = ?, last_updated = ? WHERE id = ?`,
        [valid_to.getTime(), Date.now(), id],
    );
    console.error(
        `[TEMPORAL] Invalidated fact ${id} at ${valid_to.toISOString()}`,
    );
};

export const delete_fact = async (id: string): Promise<void> => {
    await run_async(`DELETE FROM temporal_facts WHERE id = ?`, [id]);
    console.error(`[TEMPORAL] Deleted fact ${id}`);
};

export const insert_edge = async (
    source_id: string,
    target_id: string,
    relation_type: string,
    valid_from: Date = new Date(),
    weight: number = 1.0,
    metadata?: Record<string, any>,
): Promise<string> => {
    const id = randomUUID();
    const valid_from_ts = valid_from.getTime();

    await run_async(
        `
        INSERT INTO temporal_edges (id, source_id, target_id, relation_type, valid_from, valid_to, weight, metadata)
        VALUES (?, ?, ?, ?, ?, NULL, ?, ?)
    `,
        [
            id,
            source_id,
            target_id,
            relation_type,
            valid_from_ts,
            weight,
            metadata ? JSON.stringify(metadata) : null,
        ],
    );

    console.log(
        `[TEMPORAL] Created edge: ${source_id} --[${relation_type}]--> ${target_id}`,
    );
    return id;
};

export const invalidate_edge = async (
    id: string,
    valid_to: Date = new Date(),
): Promise<void> => {
    await run_async(`UPDATE temporal_edges SET valid_to = ? WHERE id = ?`, [
        valid_to.getTime(),
        id,
    ]);
    console.log(`[TEMPORAL] Invalidated edge ${id}`);
};

export const batch_insert_facts = async (
    facts: Array<{
        subject: string;
        predicate: string;
        object: string;
        valid_from?: Date;
        confidence?: number;
        metadata?: Record<string, any>;
        project_id?: string;
    }>,
    user_id?: string,
    project_id?: string,
): Promise<string[]> => {
    const ids: string[] = [];

    await run_async("BEGIN TRANSACTION");
    try {
        for (const fact of facts) {
            const id = await insert_fact(
                fact.subject,
                fact.predicate,
                fact.object,
                fact.valid_from,
                fact.confidence,
                fact.metadata,
                user_id,
                fact.project_id || project_id,
            );
            ids.push(id);
        }
        await run_async("COMMIT");
        console.log(`[TEMPORAL] Batch inserted ${ids.length} facts`);
    } catch (error) {
        await run_async("ROLLBACK");
        throw error;
    }

    return ids;
};

export const apply_confidence_decay = async (
    decay_rate: number = 0.01,
): Promise<number> => {
    const now = Date.now();
    const one_day = 86400000;

    // Postgres: use RETURNING 1 and count the rows of the result, since
    // SQLite's connection-scoped `changes()` is unavailable.
    // SQLite: run the UPDATE then read `changes()` from the same connection.
    let changes = 0;

    if (is_pg) {
        // GREATEST is the Postgres analogue of SQLite's MAX(scalar, scalar).
        const rows = await all_async(
            `
            UPDATE temporal_facts
            SET confidence = GREATEST(0.1, confidence * (1 - ? * ((? - valid_from) / ?)))
            WHERE valid_to IS NULL AND confidence > 0.1
            RETURNING 1
        `,
            [decay_rate, now, one_day],
        );
        changes = Array.isArray(rows) ? rows.length : 0;
    } else {
        await run_async(
            `
            UPDATE temporal_facts
            SET confidence = MAX(0.1, confidence * (1 - ? * ((? - valid_from) / ?)))
            WHERE valid_to IS NULL AND confidence > 0.1
        `,
            [decay_rate, now, one_day],
        );
        const result = (await get_async(`SELECT changes() as changes`)) as any;
        changes = result?.changes || 0;
    }

    console.log(`[TEMPORAL] Applied confidence decay to ${changes} facts`);
    return changes;
};

export const get_active_facts_count = async (): Promise<number> => {
    const result = (await get_async(
        `SELECT COUNT(*) as count FROM temporal_facts WHERE valid_to IS NULL`,
    )) as any;
    return result?.count || 0;
};

export const get_total_facts_count = async (): Promise<number> => {
    const result = (await get_async(
        `SELECT COUNT(*) as count FROM temporal_facts`,
    )) as any;
    return result?.count || 0;
};

/**
 * Authenticated point-lookup for a single fact. Returns null if the fact
 * either does not exist or belongs to a different tenant. This is the
 * preferred way for route handlers to confirm ownership before mutating
 * a fact, replacing the old `query_facts_at_time(...).find(id)` pattern
 * which fetched the caller's entire history just to authorize one row.
 */
export const get_fact_by_id_for_user = async (
    id: string,
    user_id: string,
    project_id?: string,
): Promise<TemporalFact | null> => {
    const params: any[] = [id, user_id];
    let sql = `SELECT id, user_id, project_id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
         FROM temporal_facts WHERE id = ? AND user_id = ?`;
    if (project_id) {
        sql +=
            " AND (project_id = ? OR project_id = 'system_global' OR project_id IS NULL)";
        params.push(project_id);
    }
    sql += " LIMIT 1";

    const row = await get_async(sql, params);
    if (!row) return null;
    return {
        id: row.id,
        user_id: row.user_id,
        project_id: row.project_id,
        subject: row.subject,
        predicate: row.predicate,
        object: row.object,
        valid_from: new Date(row.valid_from),
        valid_to: row.valid_to ? new Date(row.valid_to) : null,
        confidence: row.confidence,
        last_updated: new Date(row.last_updated),
        metadata: row.metadata ? JSON.parse(row.metadata) : undefined,
    };
};
